### DISCUSSION TITLE
PCA Visualisation

### DISCUSSION THREAD

**Discussion by rstuckey** [alan-turing-institute/sktime#2741] [Q&A]
Hi,

This may be a more general question regarding PCA on time-series data, not specific to sktime.

Is there a simple way to visualise the clusters associated with components derived from `PCATransformer`? I wish to perform dimensionality reduction on my data (to 2 or 3 dimensions) and plot the transformed data in that space to get an initial idea of the number and spread of clusters for subsequent clustering and classification. I'm not sure if plotting the _raw_ transformed data (for concatenated time-series) would make much sense, so I suspect it might have to be reduced to one data point per time-series (calculated using DTW or similar) first. Unlike **sklearn**, the `transform()` function does not return the reduced components, but the data transformed back into its original space.

I'm aware of the `plot_cluster_algorithm()` function for `TimeSeriesKMeans` and `TimeSeriesKMedoids`, but wish to plot the clusters on the same set of axes, not separately, if possible.

Thanks for any advice!

**Comment by fkiraly** [comment]
I think you may want to plot the *loadings* (= prototypical curves) and *scores* (= mixture coefficients of prototypical curves).

The loadings, for time series PCA, are time series themselves, with index support identical to the time series you carry out PCA on.
There are as many scores as time series, and they can be used for visualization in a dimension reduced space.

Currently, you can access these via the `pca` attribute of `PCATransformer` (which is an sklearn `PCA` object), though it is a bit of a hack.

I think we may like to add a function which returns the loadings as a time series panel in an `sktime` compatible format, rather than the internal numpy array.

**Reply by rstuckey** [reply]
Thanks @fkiraly !

I extracted the loadings with the attribute `pca.components_`, a numpy array, where each loading is a concatenation of the time-series features. So it has shape `(n_components, n_timepoints*n_features)`.

However, the method `pca.score()` being an **sklearn** method, requires an array-like of shape `(n_samples, n_features)` and not a panel of `(n_instances, n_features, n_timepoints)`. Do I need to transform my panel data in order to use `pca.score()` and if so, is there any point in creating a `PCATransformer` in the first place (if I am just going to use the underlying **sklearn** object anyway)?

Sorry if my terminology is incorrect. I'm still learning how to work with the framework!

**Comment by BradKML** [comment]
Also for visualizations in general you can check out https://github.com/erdogant/pca
